build scrapper passed

heytulsiprasad opened the pull request:
## Description
This PR is **WIP (work in progress)**, focused to solve #68 . The `scraper.js` built within **router** directory contains the code to fetch data from `zairza.blog.in`. It fetches `title, href, author, release-date, and cover-img-link` of all the blog posts and stores them as objects.

### Dependencies Added
* request
* cheerio


### Work remaining todo:
- [x] Build: scraper to fetch contents from `zairza.blog.in` and are stored as objects
- [x] Fix: above mentioned issue
- [x] Make this code accessible through the `app.js` file
- [x] Create a template in `blog.ejs` for rendering the blog sections
- [x] Run a forEach method on `blogs.json` at the end to fetch scraped data to our website

#### All suggestions are appreciated. 👍

heytulsiprasad committed:
add ext links to achievements

heytulsiprasad committed:
project card rotate animation

heytulsiprasad committed:
:zap: chore: build scraper for blog.ejs

scraper.js scraps from blog.zairza.in info regarding all blog posts

heytulsiprasad committed:
:sparkles: feat: build scrapper success

build a scrapper with details to fetch into blogs.ejs

heytulsiprasad committed:
:art: chore: add ejs template to blogs.ejs

heytulsiprasad committed:
:iphone: feat: make responsive layout+ add regex

layout of blogs section is made responsive and date mentioned in blog cards is rendered using regex syntax

heytulsiprasad commented:
## Feature Display
![Annotation 2019-12-21 212013](https://user-images.githubusercontent.com/52394293/71310358-46560600-2439-11ea-9026-5b78f31501db.png)

## Concerns
- [ ] background image zoom out
- [ ] blog title moved a little right (unaligned with author name)
- [x] date format yyyy/mm/dd which was originally dd/mm/yyyy

#### Help and suggestions on these above mentioned concerns is appreciated. 👍

heytulsiprasad committed:
:ok_hand: style: date format change to DD/MM/YYYY

date used in blog cards are now of the format DD/MM/YYYY

heytulsiprasad committed:
:boom: feat: fetch latest blog details

the latest post from medium is now also fetched and added to data.json file

ankitjena commented on routes/scraper.js:
```diff
@@ -0,0 +1,105 @@
+// The scraper for the blog.ejs section in the application. 
```
Where do you run this file?

ankitjena commented on routes/data.json:
```diff
@@ -0,0 +1,78 @@
+        "href": "https://blog.zairza.in/oauth-using-mevn-stack-4b4a383dae08?source=collection_home---6------0-----------------------",
+        "author": "Ramakrishna Pattnaik",
+        "release": "2019-08-25T12:13:49.122Z",
+        "cover": "https://cdn-images-1.medium.com/fit/t/1600/480/1*zqCh8ZNR-LjBzaacpiIyUA.png"
```
Since you are fetching the cover image, it's huge. Which is why it's zoomed out. We need the image which the first image inside the blog. WDYT? Can this be done?

heytulsiprasad replied:
Yes, I think so. We need to scrape every hrefs of particular blog to get the right image. Working on it now.

heytulsiprasad replied:
Mostly we'll do this step for 4 of the recent posts for optimization.

ankitjena replied:
yeah, that should reduce unnecessary requests

ankitjena commented on routes/index.js:
```diff
@@ -1,10 +1,29 @@
+var json = require('./data.json')
+
+// need 4 elements from the json array in blog.ejs file
+let data = []
```
`data` is a bad name for a variable. Always name variables according to their purpose. In this case `blogs` is better.

ankitjena commented on routes/index.js:
```diff
@@ -1,10 +1,29 @@
+for (let i=0; i<=3; i++) {
+  let release = json[i].release
+
+  let date = json[i].release.split("T")[0].replace(/-/g, '/')
```
Instead of doing this, there's a library called `momentjs` check it out.

heytulsiprasad replied:
Great package indeed, added in commit [04a095](https://github.com/zairza-cetb/zairza-website/pull/73/commits/04a095ad9b786f91340497477a0377385293d885) :tada:

ankitjena reviewed (changes requested):
@tulsi-prasad Good work, try if what I mentioned above can be done

heytulsiprasad committed:
:heavy_plus_sign: add moment as dependency

add moment js as a dependency to work with date time objects

heytulsiprasad committed:
:ok_hand: update index.js with moment pkg

write datetime objects to be rendered using moment package

heytulsiprasad committed:
:zap: build: scrapers for individual blog posts

Scrapes the first image from each blog posts and forms a cover object.

heytulsiprasad committed:
:construction: chore: add cover objects in json

added cover objects with img urls in cover.json file of first 4 blogs

heytulsiprasad commented:
## Work to Do
Fetch the cover image urls from each individual blogs (first-image). For this purpose is taken care of in [567dc39](https://github.com/zairza-cetb/zairza-website/pull/73/commits/567dc395a5cbfa4c855177965644bf23fa6d6953). And also next commit adds them to `cover.json` file for convinience.

## Problems now facing
The order of scraped image urls is not according to the blog posts order. This seems to appear out of nowhere, as the while loop iterates from `count = 0` and which fetches from `data` array which is also ordered. After fixing this only, we can render them on ejs template.

## Fix to Try
I am thinking of making a different `coverScraper.js` file which imports from `data.json` as it is and loops through its **Top 4** hrefs and fetch the first image. I'll update on this by tonight.

# EDIT 1
Refactored the code in the next commit, [2b837e5](https://github.com/zairza-cetb/zairza-website/pull/73/commits/2b837e5397e190b63e6b3d22a2f76fbc701023b6). `./json/cover.json` stores cover image urls, which are still not in order. After every surver run, the order of json file changes. Reason still doubtful.

#### Any suggestions are appreciated. 👍

heytulsiprasad committed:
:recycle: refactor: scraper codes and json

json folder stores the scraped data and scrapcover is fetches cover images from each blogs

heytulsiprasad committed:
:alembic: make an array of four urls to scrap

bloglinks array contains four urls in order to be scraped for cover image

ankitjena commented:
>I am thinking of making a different coverScraper.js file which imports from data.json as it is and loops through its Top 4 hrefs and fetch the first image

You should keep the entire logic in one file. When you fetch the blog url for cover image, do another request using cheerio. async/await will help.

heytulsiprasad commented:
This PR is taken further in #74 to avoid any local conflicts.

heytulsiprasad commented:
Have worked on a new branch. Will update in a another PR.
